### PR TITLE
ISPN-6745 Locks are lost in pessimistic cache

### PR DESCRIPTION
**Pull request by pruivo**
https://issues.jboss.org/browse/ISPN-6745

### CODE REVIEW

File: core/src/main/java/org/infinispan/interceptors/locking/PessimisticLockingInterceptor.java [hunk 1]
```diff
@@ -232,9 +232,7 @@ private boolean readNeedsLock(InvocationContext ctx, LocalFlagAffectedCommand co
    @Override
    public CompletableFuture<Void> visitApplyDeltaCommand(InvocationContext ctx, ApplyDeltaCommand command) throws Throwable {
       ctx.onReturn((rCtx, rCommand, rv, throwable) -> {
-         if (throwable != null) {
-            lockManager.unlockAll(rCtx);
-         }
+         rethrowAndReleaseLocksIfNeeded(rCtx, throwable);
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/interceptors/locking/PessimisticLockingInterceptor.java]: @pruivo Please make this change for read commands as well - `StateTransferInterceptor` also catches `OutdatedTopologyException` and retries for them, even though not not all of them can throw `OTE`.

**Comment by pruivo** [comment]
updated!

**Comment by slaskawi** [comment]
Looks good to me

**Comment by tristantarrant** [comment]
Can we have a version for 8.2.x too please ?

**Comment by pruivo** [comment]
8.2.x PR: https://github.com/infinispan/infinispan/pull/4460

**Comment by tristantarrant** [comment]
Pushed thanks

**Pull request closed without merging**